src/onnx/onnx_parser.cpp

```diff
@@ -391,18 +387,154 @@ parse_inputs(const onnx_parser& parser,
+{
+    for(size_t node_index = 0; node_index < graph.node_size(); node_index++)
+    {
+        const onnx::NodeProto& node = graph.node(node_index);
+
+        // To track actual node outputs, a future node will mention the current node's output
+        // in its list of inputs. For example:
+        // node A: outputs ["a_out"]
+        // node B: inputs ["a_out"]
+        // Here the output of node A is node B because "a_out" is the output of node A and
+        // input of node B.
+        // There can be multiple outputs, so if each new node sees an input already
+        // in the map, then append to it.
+        // In the prior example, if we add the following:
+        // node C: inputs ["a_out"]
+        // then we will have something like {"a_out": [node B, node C]} in our map.
+        for(auto&& input : node.input())
+        {
+            input_to_node_map[input].push_back(node_index);
+        }
+
+        std::vector<std::string> node_outputs;
+        std::transform(node.output().begin(),
+                       node.output().end(),
+                       std::back_inserter(node_outputs),
+                       [](const auto& output) { return output; });
+
+        // Use this second map to keep track of all references of the output names to be used
+        // as inputs to future nodes.
+        node_to_output_map.emplace(node_index, std::move(node_outputs));
+    }
+}
+
+static void traverse(std::vector<size_t>& sorted_nodes,
+                     std::unordered_set<size_t>& visited_nodes,
+                     std::map<std::string, std::vector<size_t>>& input_to_node_map,
+                     std::map<size_t, std::vector<std::string>>& node_to_output_map,
+                     size_t curr_node)
+{
+    if(contains(visited_nodes, curr_node))
+        return;
+
+    visited_nodes.insert(curr_node);
+
+    for(const auto& out_node_name : node_to_output_map.at(curr_node))
+    {
+        // check if node output is used in graph
+        if(contains(input_to_node_map, out_node_name))
+        {
+            for(const auto& in_node_name : input_to_node_map.at(out_node_name))
+                traverse(sorted_nodes,
+                         visited_nodes,
+                         input_to_node_map,
+                         node_to_output_map,
+                         in_node_name);
+        }
+    }
+    sorted_nodes.insert(sorted_nodes.begin(), curr_node);
+}
+
+static std::vector<size_t> toposort(const onnx::GraphProto& graph)
+{
+    std::map<std::string, std::vector<size_t>> input_to_node_map;
+    std::map<size_t, std::vector<std::string>> node_to_output_map;
+    create_node_maps(input_to_node_map, node_to_output_map, graph);
+
+    std::vector<size_t> sorted_nodes;
+    std::unordered_set<size_t> visited_nodes;
+
+    for(size_t node_index = 0; node_index < graph.node_size(); node_index++)
+    {
+        traverse(sorted_nodes, visited_nodes, input_to_node_map, node_to_output_map, node_index);
+    }
```

Comment:
Is this is to handle the case where we have unattached nodes?

Reply:
By unattached nodes, do you mean nodes which do not lead from a graph input or end at a graph output? This loop is needed since we traverse on node outputs. For example:

Graph input: x, y
Graph output: z
Nodes: A, B, C
x->A->C
y->B ->C
A,B->C->z

Here if we only traversed starting from A, we would be missing node B since it does not depend on A.

If using my definition of unattached nodes above, then this loop would still iterate through them, but dead code elimination will eventually get rid of them anyways.